doc/code-tests.md

```diff
@@ -0,0 +1,190 @@
+tests are organized by source directory. Hence, if the failure is occurring in
+a test associated with the `src/shared` directory (say), then you can run just
+the tests associated with that directory via `npm run test:unit src/shared`.
+For a brief overview of other options you might use with running unit tests,
+try `npm run test:unit -- --help`. (Note the extra `--` stuck in there, which
+is needed because the test option you want to use here itself begins with
+`--`.)
+
+#### Customizing Playwright runs
+
+You may find yourself in a similar situation with end-to-end testing. In this
+case, all of the tests are specified in the top-level `e2e` directory.
+However, they are organized into thematic files, which you can see via
+`ls e2e`, for example. You can run just one of these collections of tests, say
+the ones for the Gallery view in `e2e/gallery.spec.ts`, by executing
+`npm run test:e2e gallery`. And similarly to unit tests, you can obtain an
+overview of other possibilities via `npm run test:e2e -- --help`.
+
+## Creating tests
+
+If you are considering contributing code to Numberscope, you will very likely
+need to create tests. The [pull request checklist](pull-request-checklist.md)
+notes that any new feature (including a new visualizer) requires some test
+that exercises it; and any bug fix should include a test that would have
+failed with the buggy code but which now succeeds on the improved code. There
+are somewhat different processes for adding each of the two kinds of tests.
+
+### Adding a unit test
+
+As mentioned above, these tests are organized per source directory, in
+subdirectories named `__tests__`. More than likely such a directory will
+already exist alongside the source file you want to test, but if not, simply
+create one -- vitest will find it automatically. Within the **tests**
+directory, the tests are organized by source file, and named accordingly, with
+file extension `.spec.ts`. Again, if you are adding tests related to a file
+that already has associated tests, you can just insert new tests in an
+existing file. If there are no tests yet for the file you are concerned with,
+you can just create the `.spec.ts` file and the framework will automatically
+execute it.
+
+Let's go through the latter case. At the time of this writing, there was no
+test specification for the `src/shared/defineFeatured.ts` file that provides
+the specimens for the Featured Gallery. So, we simply created
+`src/shared/__tests__/defineFeatured.spec.ts`. At the top of the file you need
+to import the test facilities from `vitest`, and typically all of the exported
+entities from the module being tested, like so:
+
+```
+{! ../src/shared/__tests__/defineFeatured.spec.ts extract: { stop: ibe..g } !}
+```
+
+In this case, there is only one export, `getFeatured`, so the test
+specification is pretty brief. Here it is in full:
+
+```
+{! ../src/shared/__tests__/defineFeatured.spec.ts extract: { start: tured } !}
+```
```

Comment:
Is this correct?  what's going on with these two code lines?  They aren't in the defineFeatured.spec.ts file.  Are these supposed to pull some code automatically from that file?  Right now when I view this md file in the pull request on github that doesn't work.  Is there some context in which it works?  It would need to work on github.com once the PR is merged, since people read these md files there.

Reply:
From my perspective, since we started using the "mkdocs" documentation system, the documentation files are simply the source code (as it were) for the static documentation site produced by mkdocs and installed along with the numberscope app and reachable via the "Documentation" link on the numberscope site. You can preview just these docs and not the app but with hot-reload when you change a doc file using `npm run doc:serve`. Or you can preview the entire site, app and docs, but without hot-reload, using `npm run preview`. I would suggest the doc files should be reviewed by one of those methods.

So indeed, when the markdown files make use of these inclusion (or extraction, depending on how you want to term it) directives, the direct markdown file read either as text or viewed as markdown in GitHub's automatic viewer when you look at an `.md` file is not going to show the _results_ of that inclusion -- in other words, it won't look the way it will look on the documentation site as it is installed, and it may not seem entirely sensible at those points.

I don't personally see this as a particular problem, in that I think the source for numberscope documentation, and the place we should link to for that purpose, is numberscope.colorado.edu/doc; I don't think we should be encouraging people to use the code browser in the github repository as the way to access the documentation.

If, however, you feel it is critical that the documentation read perfectly sensibly in GitHub's automatic markdown rendering in its code browser, then by introducing some noticeable complexity in each of these places where the inclusion mechanism is used, we could manage that: Basically, we would put in special markdown comments in these locations so that from the point of view of the GitHub markdown renderer, the inclusion syntax would be inside a comment and hence not be rendered; and conversely, we would put HTML links in the .md files that would automatically removed by the MkDocs processing, while getting MkDocs to respect the inclusion syntax, so that in the GitHub rendering a link to the included material would be shown (rather than the included material itself, as would be shown on the Numberscope documentation pages). So, the upside is that the documentation pages would read sensibly on both the GitHub code browser and the Numberscope site (albeit only links would show in the code browser, rather than the included material itself); and the downside is that the setup and the documentation `.md` files themselves would be noticeably more intricate and hence somewhat harder to write and maintain. Thoughts about which way to go here, @katestange / @Vectornaut ?

Reply:
OK, as discussed in meeting, the issue #460 that details how we might make the markdown files sensible both in GitHub and in the docsite is filed. So if it looks OK to you, perhaps you can resolve this conversation here.